### DISCUSSION TITLE
Convert byte array to int and float

### DISCUSSION THREAD

**Discussion by ValentinOsvaldo** [kenjdavidson/react-native-bluetooth-classic#207] [Q&A] [answered]
Currently in the company where I work we are creating a tool that sends data as bytes, what I have to do is transform that data to an integer and float type. I was looking in the library but I didn't find anything like that.

This is the code with which I connect the device

```js
const connect = await RNBluetoothClassic.connectToDevice(
          device.address,
          {
            connectorType: 'rfcomm',
            charset: 'ascii',
            delimiter: '\n',
            connectionType: 'binary',
          },
        );
```

This is what the device sends me:
![image](https://user-images.githubusercontent.com/69062379/196595568-331eeace-95e1-452e-b8c7-0e095fd018e4.png)

Unfortunately I still can't find the documentation for the device, but the hardware team sent me the following code:

![image](https://user-images.githubusercontent.com/69062379/196595084-36c66463-c6b4-44de-8d5c-39f34a76b112.png)

If there is any other information that I can provide, I will gladly try

**Comment by kenjdavidson** [comment, marked as answer]
React native has a limitation on what data can be transferred between Javascript and native bridge. The connection defaults to string, so you're always going to need to keep this in mind.

Rfcomm is good
Delimiter does nothing
Charset does nothing
All due to binary using byte array connection

On read the byte array is being converted to a string with this

```
Base64.encodeToString(arr, Base64.DEFAULT);
```

So on the Javascript side you need to reverse this into a byte array and then I to the ints and floats based on the communication protocol.

If you need, you can always write a custom connection that performs the conversion on Java and then send ints and floats to Javascript as strings to just parse.

Dealers Choice.

**Reply by ValentinOsvaldo** [reply]
# Convert base64 to bufferArray

I'm happy to give this answer, even if it took a while, I hope this helps more people

So make the connection with the device:
```ts
const connect = await RNBluetoothClassic.connectToDevice(
  device.address,
  {
    connectorType: 'rfcomm',
    connectionType: 'binary',
  },
);
```

I use the [buffer](https://www.npmjs.com/package/buffer) package

```ts
import { Buffer } from 'buffer'
```

```ts
device.onDataReceived(({ data }) => {
  const buffer = Buffer.from(data, 'base64');

  const arrayBuffer = buffer.buffer.slice(
    buffer.byteOffset, buffer.byteOffset + buffer.byteLength
  )

  if (arrayBuffer.byteLength === 4) {
    const rimDepth = new Float32Array(new Uint8Array(arrayBuffer).buffer)[0].toFixed(2);
    return;
  }
 ...
});
```

The results look like this:
![image](https://user-images.githubusercontent.com/69062379/198850286-2aa92180-d61b-46c0-8648-9e39267068c4.png)

**Reply by scalhoun** [reply]
Thank you very much for sharing this!

**Comment by kenjdavidson** [comment]
It looks like the protocol is 5 bytes

Byte 1 is the type
Byte 2 to 5 is the value

There is no delimiation in the default logic so let's assume that each message is a byte.

So 

OnRead I'll give you a base64 string
Btoa(value) give you the array
Now just convert the Java code to Javascript. 

Assuming you're doing the receiving.

**Comment by kenjdavidson** [comment]
Solid. If you want to add to the documentation site I'll accept a pr.
